File: sdk/node/src/stream.ts
/**
 * SSE stream reader. Parses Server-Sent Events from a Response body into an
 * async iterable of parsed JSON chunks.
 *
 * The Wingman proxy streams OpenAI-style chunks via SSE: each event is a
 * single `data:` line whose payload is either a JSON object
 * (`{"choices":[{"delta":{"content":"hi"}}],...}`) or the literal `[DONE]`
 * sentinel terminating the stream.
 */

import { APIError, WingmanError } from "./errors.js";

export interface ChatStreamChunk {
  id?: string;
  object?: string;
  created?: number;
  model?: string;
  choices?: Array<{
    index?: number;
    delta?: { role?: string; content?: string };
    finish_reason?: string | null;
  }>;
  [key: string]: unknown;
}

export class Stream<Item> implements AsyncIterable<Item> {
  readonly controller: AbortController;
  #iteratorFactory: () => AsyncIterator<Item>;
  #consumed = false;

  constructor(iteratorFactory: () => AsyncIterator<Item>, controller: AbortController) {
    this.#iteratorFactory = iteratorFactory;
    this.controller = controller;
  }

  static fromSSEResponse<Item = ChatStreamChunk>(
    response: Response,
    controller: AbortController
  ): Stream<Item> {
    if (!response.body) {
      throw new WingmanError("Streaming response has no body");
    }

    let consumed = false;
    const body = response.body;

    async function* iterator(): AsyncIterator<Item> {
      if (consumed) {
        throw new WingmanError(
          "Cannot iterate over a consumed stream; create a new request instead."
        );
      }
      consumed = true;

      const reader = body.getReader();
      const decoder = new TextDecoder("utf-8");
      let buffer = "";

      try {
        while (true) {
          const { value, done } = await reader.read();
          if (done) break;
          buffer += decoder.decode(value, { stream: true });

          // SSE frames are separated by a blank line ("\n\n").
          let sep: number;
          while ((sep = buffer.indexOf("\n\n")) !== -1) {
            const frame = buffer.slice(0, sep);
            buffer = buffer.slice(sep + 2);

            let eventName: string | null = null;
            const dataLines: string[] = [];
            for (const rawLine of frame.split("\n")) {
              const line = rawLine.replace(/\r$/, "");
              if (!line || line.startsWith(":")) continue;
              const colon = line.indexOf(":");
              const field = colon === -1 ? line : line.slice(0, colon);
              const value =
                colon === -1
                  ? ""
                  : line.slice(colon + 1).replace(/^ /, "");
              if (field === "event") eventName = value;
              else if (field === "data") dataLines.push(value);
            }

            if (dataLines.length === 0) continue;
            const data = dataLines.join("\n");
            if (data === "[DONE]") return;

            if (eventName === "error") {
              let parsed: unknown;
              try {
                parsed = JSON.parse(data);
              } catch {
                parsed = data;
              }
              throw APIError.generate(
                undefined,
                typeof parsed === "object" ? (parsed as Record<string, unknown>) : data,
                "Stream error",
                response.headers
              );
            }

            try {
              yield JSON.parse(data) as Item;
            } catch {
              // Non-JSON chunk (rare). Surface as a raw string.
              yield data as unknown as Item;
            }
          }
        }
      } finally {
        try {
          reader.releaseLock();
        } catch {
          /* ignore */
        }
      }
    }

    return new Stream<Item>(iterator, controller);
  }

  [Symbol.asyncIterator](): AsyncIterator<Item> {
    if (this.#consumed) {
      throw new WingmanError(
        "Stream already consumed; issue a new request to stream again."
      );
    }
    this.#consumed = true;
    return this.#iteratorFactory();
  }

  /** Abort the underlying fetch and stop iteration. */
  abort(): void {
    this.controller.abort();
  }
}

File: sdk/python/src/wingman/stream.py
"""SSE stream parser for Wingman chat responses."""

from __future__ import annotations

import json
from typing import Any, Dict, Iterator, Mapping, Optional

import httpx

from .errors import APIError, WingmanError


ChatStreamChunk = Dict[str, Any]


class Stream:
    """Iterable wrapper over an SSE response body.

    Yields parsed JSON objects (one per SSE frame). The `[DONE]` sentinel
    terminates iteration. `event: error` frames raise an `APIError`.
    """

    def __init__(self, response: httpx.Response) -> None:
        self._response = response
        self._consumed = False

    def __iter__(self) -> Iterator[ChatStreamChunk]:
        if self._consumed:
            raise WingmanError(
                "Cannot iterate a consumed stream; issue a new request to stream again."
            )
        self._consumed = True

        try:
            buffer = ""
            for raw in self._response.iter_text():
                if not raw:
                    continue
                buffer += raw
                while True:
                    sep = buffer.find("\n\n")
                    if sep == -1:
                        break
                    frame = buffer[:sep]
                    buffer = buffer[sep + 2 :]
                    event = self._parse_frame(frame)
                    if event is None:
                        continue
                    event_name, data = event
                    if data == "[DONE]":
                        return
                    if event_name == "error":
                        try:
                            parsed: Any = json.loads(data)
                        except json.JSONDecodeError:
                            parsed = data
                        raise APIError.generate(
                            None,
                            parsed if isinstance(parsed, Mapping) else data,
                            "Stream error",
                            dict(self._response.headers),
                        )
                    try:
                        yield json.loads(data)
                    except json.JSONDecodeError:
                        yield {"content": data}  # fallback
        finally:
            self.close()

    @staticmethod
    def _parse_frame(frame: str) -> Optional[tuple[Optional[str], str]]:
        event_name: Optional[str] = None
        data_lines: list[str] = []
        for raw_line in frame.split("\n"):
            line = raw_line.rstrip("\r")
            if not line or line.startswith(":"):
                continue
            colon = line.find(":")
            if colon == -1:
                field, value = line, ""
            else:
                field = line[:colon]
                value = line[colon + 1 :]
                if value.startswith(" "):
                    value = value[1:]
            if field == "event":
                event_name = value
            elif field == "data":
                data_lines.append(value)
        if not data_lines:
            return None
        return event_name, "\n".join(data_lines)

    def text_deltas(self) -> Iterator[str]:
        """Yield only the incremental text deltas from each chunk."""
        for chunk in self:
            delta = (
                (chunk.get("choices") or [{}])[0].get("delta", {}).get("content")
                if isinstance(chunk.get("choices"), list)
                else None
            )
            if delta:
                yield delta

    def close(self) -> None:
        try:
            self._response.close()
        except Exception:
            pass

    def __enter__(self) -> "Stream":
        return self

    def __exit__(self, *_exc: Any) -> None:
        self.close()
